Refine Plugin Framework Docs

laurapacilio opened the pull request:
Continuing to refine the docs for the new plugin framework SDK:

**Add an SDK homepage.**
I added an overview page for the plugin SDKv2. This is for a few reasons:
- The framework has an overview page, and it makes sense to do this for consistency. Also, now that there are two ways to develop plugins, we want to make it very clear what each of these things are and when folks should be using them. 
- Fix the fact that when you click plugin development, the SDK dropdown is automatically open. 
- Fix the fact that when you click SDKv2, it shows the plugin development intro page. That would confuse me as a user, when what I'm likely looking for is information specific to SDKv2.

**Update Framework overview page**
I expanded the framework overview page a bit to hopefully help users understand what they should do next (especially if they're new to plugin development and don't yet have a good mental model). We can continue to iterate on this, but this is a start. 

**Adjust Attribute page formatting**
I used tables for the methods in the interfaces for creating your own attribute types and values. I think this makes the page easier to read because there aren't so many tiny little sections with subtitles. It also helps make the page shorter and less overwhelming. :-)  I also did a couple of minor tweaks (editing some subheads, etc.) to make them more concise.

**Adjust Publishing to the Registry introduction**
I reworked the intro section to try to make it more active voice and concise. I also added numbered steps under "Add Manifest" to hopefully make it clearer that there are two things users need to do there. 

**Update Access State, Config, and Plan title**
I changed the Access State, Config and Plan title to match the ‘ing’ we use in the sidebar. It's jarring for users when they click a page link and the name of the page doesn't match what they clicked.

**Update learn tutorial links in sidebar**
I adjusted the link names so it's hopefully clearer where those links are taking you and what you can expect from them. For new users, I don't know that it would be immediately clear that these are taking you to hands-on tutorials, so I put Tutorial in the front to help make that distinction. 

**Update Plugin Homepage**
I added a pointer to new learn tutorial for framework! :-)

bflad commented:
Submitted https://github.com/hashicorp/terraform-website/pull/1793 to help resolve the new golang.org/pkg redirects.

bflad commented:
The link checker should be much happier now if this is rebased. 😄

laurapacilio commented:
It does indeed! Thank you, @bflad!

paddycarver commented on content/source/docs/extend/sdkv2-intro.html.md:
```diff
@@ -0,0 +1,27 @@
+layout: "extend"
+page_title: "Home - Plugin Development: SDKv2"
+description: |-
+  Learn about a framework you can use to develop Terraform plugins.
```
```suggestion
  Learn about version 2 of the Terraform Plugin SDK.
```

paddycarver commented on content/source/docs/extend/sdkv2-intro.html.md:
```diff
@@ -0,0 +1,27 @@
+
+# Terraform Plugin SDKv2
+
+Terraform Plugin SDKv2 is a precursor to the [Terraform Plugin Framework](/docs/plugin/framework/index.html) for developing Terraform plugins.
+
+~> **Important**: If you maintain a large existing provider, we recommend that you continue using SDKv2 to develop new resources and data sources. [Which SDK Should I Use?](/docs/plugin/which-sdk.html) explains the differences between SDKv2 and the framework.
```
I think this is more aggressive than we'd like to be at this time. @aareet and @redeux can speak with more authority on this, but I think we're trying to position SDKv2 as the main SDK at the moment, with the framework being the early-adopter next version until the framework is more capable.

laurapacilio replied:
Noted! Check the updated file and see if you think that's better :)

paddycarver commented on content/source/docs/plugin/framework/types.html.md:
```diff
@@ -233,93 +233,54 @@ method](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/types
 State, Config, and Plan](/docs/plugin/framework/accessing-values.html).
 
-## Create Your Own Types and Values
+## Create Types and Values
```
```suggestion
## Create Provider-Defined Types and Values
```

paddycarver reviewed (commented):
This is a really nice update. Some minor comments.

laurapacilio commented on content/source/docs/extend/sdkv2-intro.html.md:
```diff
@@ -0,0 +1,26 @@
+
+# Terraform Plugin SDKv2
+
+Terraform Plugin SDKv2 is an established way to develop Terraform Plugins. It is the precursor to the [Terraform Plugin Framework](/docs/plugin/framework/index.html), which is still in early development.
```
@paddycarver Noted! I updated this a bit to try to get closer to what you described - the SDKv2 is still established and in many cases preferred, but that the framework is coming! and then I opted away from making a specific recommendation and just point folks toward the which SDK page to learn more.

paddycarver replied:
As long as @aareet or @redeux is on board with this framing, it works for me.

laurapacilio replied:
Thank you! @aareet I tagged you for review - let me know what you think when you get a chance!

aareet replied:
Looks good to me, thank you for writing all this up!

laurapacilio commented on content/source/docs/plugin/which-sdk.html.md:
```diff
@@ -8,17 +8,12 @@ description: |-
-
-This guide will help you decide whether you should stick with writing providers
-in SDKv2, or prepare for the future by writing providers in the framework.
+There are multiple SDKs that you can use to build Terraform providers:
```
@paddycarver I updated this too - just formatting to make it easier to read and made it a bit shorter.

laurapacilio commented:
Thank you for the review @paddycarver! I accepted your edits and addressed your comments - see if you think this is better with the new copy. Thank you!

paddycarver commented on content/source/docs/extend/sdkv2-intro.html.md:
```diff
@@ -0,0 +1,26 @@
+
+## Get Started
+
+Try the [Call APIs with Custom Providers](https://learn.hashicorp.com/collections/terraform/providers?utm_source=WEBSITE&utm_medium=WEB_IO&utm_offer=ARTICLE_PAGE&utm_content=DOCS) tutorials on Hashicorp Learn.
```
```suggestion
Try the [Call APIs with Custom Providers](https://learn.hashicorp.com/collections/terraform/providers?utm_source=WEBSITE&utm_medium=WEB_IO&utm_offer=ARTICLE_PAGE&utm_content=DOCS) tutorials on HashiCorp Learn.
```

paddycarver reviewed (approved):
Minor suggestions, and I'm punting to @aareet or @redeux on the framing of SDKv2 vs framework, but this seems good to me. 👍 great work!

aareet commented on content/source/docs/extend/index.html.md:
```diff
@@ -18,10 +18,10 @@ Terraform currently supports one type of Plugin called [providers](/docs/languag
 ## Develop and Share Providers
-- Learn more about **how to use SDKv2** to develop providers. This includes details about how to define attributes and behaviors using [schemas](/docs/extend/schemas/index.html), [develop resources](/docs/extend/resources/index.html), [debug providers](/docs/extend/debugging.html), and [test plugins](/docs/extend/testing/index.html).
+- Learn more about **how to use [SDKv2](/docs/extend/sdkv2-intro.html)** to develop providers. This includes details about how to define attributes and behaviors using [schemas](/docs/extend/schemas/index.html), [develop resources](/docs/extend/resources/index.html), [debug providers](/docs/extend/debugging.html), and [test plugins](/docs/extend/testing/index.html).
 - Learn more about **how to use the [framework](/docs/plugin/framework/index.html)** to develop providers. This includes details about how to define attributes and behaviors using [schemas](/docs/plugin/framework/schemas.html), develop [resources](/docs/plugin/framework/resources.html) and [data sources](/docs/plugin/framework/data-sources.html), and [write tests](/docs/plugin/framework/acctests. ...
```
```suggestion
- Learn more about **how to use the [plugin framework](/docs/plugin/framework/index.html)** to develop providers. This includes details about how to define attributes and behaviors using [schemas](/docs/plugin/framework/schemas.html), develop [resources](/docs/plugin/framework/resources.html) and [data sources](/docs/plugin/framework/data-sources.html), and [write tests](/docs/plugin/framework/acctests.html).
```

aareet commented on content/source/docs/plugin/framework/publishing.html.md:
````diff
@@ -8,35 +8,26 @@ description: |-
-```
+    ```
 {
   "version": 1,
````
I suggest using a realistic example for version, like 1.0.0

aareet replied:
I would also include the file name in the example block to show how the contents are reflected in the name

aareet replied:
@paddycarver is the `version` in this file different from the $VERSION used in the file name?

laurapacilio replied:
I just pushed a change to make this "1.0.0" so that it's more realistic and matches the example for the name above. I also added text to remind folks that they shouldn't just paste it as is. @aareet What do you think?

paddycarver replied:
Version 1 is correct. This is a different version than the provider version, it's the version of the manifest format.

laurapacilio replied:
Thanks @paddycarver! @aareet I pushed one final edit that puts the version number BACK to 1, per Paddy's explanation. But I did add a little explanation in the second step that this is the version of the manifest, not the version of the provider - hopefully this way folks won't get confused. Let me know if you feel comfortable approving after this change :-)

aareet commented on content/source/docs/plugin/which-sdk.html.md:
```diff
@@ -8,17 +8,12 @@ description: |-
-
-This guide will help you decide whether you should stick with writing providers
-in SDKv2, or prepare for the future by writing providers in the framework.
+There are multiple SDKs that you can use to build Terraform providers:
```
Should we say `two SDKs` rather than `multiple`?

laurapacilio replied:
thank you! adjusted this as well!

aareet commented on content/source/docs/plugin/which-sdk.html.md:
```diff
@@ -8,17 +8,12 @@ description: |-
+There are multiple SDKs that you can use to build Terraform providers:
+
+-  [SDKv2](/docs/extend/sdkv2-intro.html) is what the majority of existing providers are built on, and it continues to provide a stable development experience.
+- [Terraform Plugin Framework](/docs/plugin/framework/index.html) is a new SDK under heavy development, realigning the provider development experience and abstractions around the updated architecture of Terraform. It represents our vision of the future of Terraform plugin development.
```
suggest `active` rather than `heavy` development

laurapacilio replied:
thank you! I pushed a change to make it "active" rather than "heavy"

aareet commented on content/source/docs/plugin/framework/publishing.html.md:
```diff
@@ -8,35 +8,26 @@ description: |-
-only supports version 6 of the Terraform protocol, and will correctly advertise
-that fact to Terraform, so Terraform versions that don't support protocol 6
-will not download it.
+The Terraform Registry will detect this file and understand that this version of your provider only supports version 6 of the Terraform protocol. It and will correctly advertise that fact to Terraform, so that Terraform versions that don't support protocol 6 will not download it.
```
```suggestion
The Terraform Registry will detect this file and understand that this version of your provider only supports version 6 of the Terraform protocol. It will correctly advertise that fact to Terraform, so that Terraform versions that don't support protocol 6 will not download it.
```

aareet reviewed (changes requested):
Made some minor comments. Otherwise the framing looks good to me

laurapacilio commented:
Also, while I was waiting I made a picture! @im2nguyen @paddycarver @aareet What do you think?!?  It's very simple, but I think it helps break up the text-heavy page a bit.  Okay, this new pic + the edit I made above to versions are my last things - if y'all approve, I'll merge!

aareet reviewed (approved):
Looks good to me - thanks for the work here Laura!
